import { Request, Response } from "express";
import { CustomError, ProvinceDto } from "../../domain";
import { ProvinceService } from "../services/province.service";

export class ProvinceController {

    protected provinceService: ProvinceService = new ProvinceService();

    private handleError(error: unknown, res: Response) {
        if (error instanceof CustomError) {
            return res.status(error.statusCode).json({ message: error.message });
        }
        console.log(error);
        res.status(500).json({ message: 'Internal server error' });
    }

    // MÃ©todos de la clase
    getAll = async (req: Request, res: Response) => {
        this.provinceService.getProvinces()
            .then((data) => {
                res.json(data);
            })
            .catch((error) => {
                this.handleError(error, res);
            });
    }

    getById = async (req: Request, res: Response) => {
        const id = req.params.id;
        if (!id) return res.status(400).json({ message: 'Missing id' });

        this.provinceService.getProvince(parseInt(id))
            .then((data) => {
                res.json(data);
            })
            .catch((error) => {
                this.handleError(error, res);
            });
    }

    create = async (req: Request, res: Response) => {
        const [error, createDto] = ProvinceDto.create(req.body);
        if (error) return res.status(400).json({ message: error });

        this.provinceService.createProvince(createDto!)
            .then((data) => {
                res.json(data);
            })
            .catch((error) => {
                this.handleError(error, res);
            });
    }

    update = async (req: Request, res: Response) => {
        const id = parseInt(req.params.id);
        if (!id) return res.status(400).json({ message: 'Missing id' });
        const [error, updateDto] = ProvinceDto.create(req.body);
        if (error) return res.status(400).json({ message: error });

        this.provinceService.updateProvince(id, updateDto!)
            .then((data) => {
                res.json(data);
            })
            .catch((error) => {
                this.handleError(error, res);
            });
    }

    delete = async (req: Request, res: Response) => {
        const id = req.params.id;
        if (!id) return res.status(400).json({ message: 'Missing id' });

        this.provinceService.deleteProvince(parseInt(id))
            .then((data) => {
                res.json(data);
            })
            .catch((error) => {
                this.handleError(error, res);
            });
    }

}